pkg/controller/portworxdiag/rbac.go

```diff
@@ -0,0 +1,121 @@
+package portworxdiag
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceAccountName = "portworx-diag-collector"
+	roleName           = "diag-collector"
+	roleBindingName    = "diag-collector"
+)
+
+func objectExists(c client.Client, ref types.NamespacedName, emptyObj client.Object) (bool, error) {
+	err := c.Get(context.TODO(), ref, emptyObj)
+	if err == nil {
+		return true, nil
+	}
+	if errors.IsNotFound(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func ensureRBACExists(client client.Client, ns string) error {
+	if err := createServiceAccountIfNotExist(client, ns); err != nil {
+		return err
+	}
+	if err := createRoleIfNotExist(client, ns); err != nil {
+		return err
+	}
+	if err := createRoleBindingIfNotExist(client, ns); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createServiceAccountIfNotExist(c client.Client, ns string) error {
+	exists, err := objectExists(c, types.NamespacedName{Namespace: ns, Name: serviceAccountName}, &corev1.ServiceAccount{})
+	if err != nil {
+		return fmt.Errorf("failed to check if service account %s/%s exists: %v", ns, serviceAccountName, err)
+	}
+	if exists {
+		return nil
+	}
+
+	if err = c.Create(context.TODO(), &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{Name: serviceAccountName, Namespace: ns},
+	}); err != nil {
+		return fmt.Errorf("failed to create service account %s/%s: %v", ns, serviceAccountName, err)
+	}
+
+	return nil
+}
+
+func createRoleIfNotExist(c client.Client, ns string) error {
+	exists, err := objectExists(c, types.NamespacedName{Namespace: ns, Name: roleName}, &rbacv1.Role{})
+	if err != nil {
+		return fmt.Errorf("failed to check if role %s/%s exists: %v", ns, roleName, err)
+	}
+	if exists {
+		return nil
+	}
+
+	if err = c.Create(context.TODO(), &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{Name: roleName, Namespace: ns},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{"portworx.io"},
+				Resources: []string{"portworxdiags", "portworxdiags/status"},
+				Verbs:     []string{"*"},
+			},
+			{
+				APIGroups: []string{"core.libopenstorage.org"},
+				Resources: []string{"storageclusters"},
```

Comment:
Don't you need to storagenode objects as well? We are listing storagenods if the rpc is not available.
Let's just put * here instead of listing all the objects under `core.libopenstorage.org`.